### PR TITLE
feat(conditional-approval): Users want conditional triggers of conditional approval

### PR DESCRIPTION
**Pull request by matthv**
### General

- [x] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [x] Test manually the implemented changes
- [x] Validate the code quality (indentation, syntax, style, simplicity, readability)

### Security

- [x] Consider the security impact of the changes made

### CODE REVIEW

File: app/controllers/forest_liana/resources_controller.rb
Comment (Thenkei) [inline app/controllers/forest_liana/resources_controller.rb:46]: My god! I love this! 👏

File: app/controllers/forest_liana/application_controller.rb
Comment (Thenkei) [inline app/controllers/forest_liana/application_controller.rb:104]: Is there a way to make a global util to do this?

File: app/controllers/forest_liana/smart_actions_controller.rb
Comment (Thenkei) [inline app/controllers/forest_liana/smart_actions_controller.rb:20]: You've not added the `signedApprovalRequest` support yet?

File: app/services/forest_liana/ability/permission.rb
Comment (Thenkei) [inline app/services/forest_liana/ability/permission.rb:18]: Clean !

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (Thenkei) [inline app/services/forest_liana/ability/permission/smart_action_approval.rb:6]: 🥇 I didn't see that coming. So in here, you should use the `signedApprovalRequest` as source of truth.

Comment (Thenkei) [reply app/services/forest_liana/ability/permission/smart_action_approval.rb:6]: Are we only handling approvals in here? (I'm a little bit lost)

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (Thenkei) [inline app/services/forest_liana/ability/permission/smart_action_approval.rb:15]: Is it the definition of a trigger or an approve?

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (Thenkei) [inline app/services/forest_liana/ability/permission/smart_action_approval.rb:19]: I must go buy some 👓

Comment (Thenkei) [reply app/services/forest_liana/ability/permission/smart_action_approval.rb:19]: But how does it replace the @parameters[:data][:attributes] by the one inside the `signed_approval_request`?

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (Thenkei) [inline app/services/forest_liana/ability/permission/smart_action_approval.rb:63]:
```suggestion
      def verifyAndDecodeSignedApprovalRequest
        decode_parameters = JWT.decode(@parameters[:data][:attributes][:signed_approval_request], ForestLiana.env_secret, true, { algorithm: 'HS256' }).try(:first)
        @parameters = ActionController::Parameters.new(decode_parameters)
      end
```
I didn't get the `deserialize` method name. I feel `verifyAndDecodeSignedApprovalRequest` more accurate.  What do you think?

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (Thenkei) [inline app/services/forest_liana/ability/permission/smart_action_approval.rb:15]:
```suggestion
      def can_execute?
```
As it checks trigger and approve?

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (Thenkei) [inline app/services/forest_liana/ability/permission/smart_action_approval.rb:29]: I think you need to put this condition first. As an approver can also have the right to trigger but in case of an approval (`signed_approval_request` param present) we want to check approval rights.

```
if @parameters[:data][:attributes][:signed_approval_request].present? && @smart_action['userApprovalEnabled'].include?(@user['roleId'])
```

File: app/controllers/forest_liana/application_controller.rb
Comment (matthv) [reply app/controllers/forest_liana/application_controller.rb:104]: This is the purpose of this method 😅. It's already available for each controller.

File: app/controllers/forest_liana/smart_actions_controller.rb
Comment (matthv) [reply app/controllers/forest_liana/smart_actions_controller.rb:20]: This is just a simple controller check, the `signedApprovalResquest` support is used on a specific module.

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (matthv) [reply app/services/forest_liana/ability/permission/smart_action_approval.rb:6]: We handle all the kind of approvals/trigger permission of a smartaction here.

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (matthv) [reply app/services/forest_liana/ability/permission/smart_action_approval.rb:15]: good idea 👍

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (matthv) [reply app/services/forest_liana/ability/permission/smart_action_approval.rb:19]: We check if the `@parameters[:data][:attributes][:signed_approval_request]` exist and at line 20 we call the `deserialize` method to replace de parameters.

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (matthv) [reply app/services/forest_liana/ability/permission/smart_action_approval.rb:63]: not bad 👍 but I think just `decodeSignedApprovalRequest` is more efficient

File: app/services/forest_liana/ability/permission/smart_action_approval.rb
Comment (matthv) [reply app/services/forest_liana/ability/permission/smart_action_approval.rb:29]: 🤔 to discuss

File: app/services/forest_liana/ability/permission/request_permission.rb
Comment (Thenkei) [inline app/services/forest_liana/ability/permission/request_permission.rb:9]:
```suggestion
          if (params[:data][:attributes][:signed_approval_request])
             if (params[:data][:attributes][:requester_id])
               raise ForestLiana::UnprocessableError.new
             end
            
            decode_parameters = JWT.decode(params[:data][:attributes][:signed_approval_request], ForestLiana.env_secret, true, { algorithm: 'HS256' }).try(:first)
```

File: app/controllers/forest_liana/smart_actions_controller.rb
Comment (Thenkei) [inline app/controllers/forest_liana/smart_actions_controller.rb:55]: Okay. You pass the input params so forest_authorize re-decode (decodeSignedApprovalRequest). Looks pretty good to me.

File: app/controllers/forest_liana/smart_actions_controller.rb
Comment (Thenkei) [inline app/controllers/forest_liana/smart_actions_controller.rb:5]: 😍

**Review by nicolasalexandre9 — APPROVED** [review]

**Pull request merged**
